x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,170 @@
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return updateTimeSeriesTemporalRange(currentState, Instant.now());
+            }
+
+            @Override
+            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                onComplete.run();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                LOGGER.warn("failed to update tsdb data stream end times", e);
+                onComplete.run();
+            }
+
+        }, ClusterStateTaskExecutor.unbatched());
+    }
+
+    void setPollInterval(TimeValue newValue) {
+        LOGGER.info(
+            "updating ["
+                + DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey()
+                + "] setting from ["
+                + pollInterval
+                + "] to ["
+                + newValue
+                + "]"
+        );
+        this.pollInterval = newValue;
+    }
+
+    ClusterState updateTimeSeriesTemporalRange(ClusterState current, Instant now) {
+        Metadata.Builder mBuilder = null;
+        for (DataStream dataStream : current.metadata().dataStreams().values()) {
+            if (dataStream.getIndexMode() != IndexMode.TIME_SERIES) {
+                continue;
+            }
+            if (dataStream.isReplicated()) {
+                continue;
+            }
+
+            // getWriteIndex() selects the latest added index:
+            String head = dataStream.getWriteIndex().getName();
+            IndexMetadata im = current.metadata().getIndices().get(head);
+            Instant currentEnd = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+            TimeValue lookAheadTime = IndexSettings.LOOK_AHEAD_TIME.get(im.getSettings());
+            Instant newEnd = now.plus(lookAheadTime.getMillis(), ChronoUnit.MILLIS).plus(pollInterval.getMillis(), ChronoUnit.MILLIS);
+            if (newEnd.isAfter(currentEnd)) {
+                Settings settings = Settings.builder()
+                    .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), DEFAULT_DATE_TIME_FORMATTER.format(newEnd))
+                    .build();
+                LOGGER.debug(
+                    "updating [{}] setting from [{}] to [{}] for index [{}]",
+                    IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                    currentEnd,
+                    newEnd,
+                    head
+                );
+                if (mBuilder == null) {
+                    mBuilder = Metadata.builder(current.metadata());
+                }
+                mBuilder.updateSettings(settings, head);
+                Metadata.Builder finalMBuilder = mBuilder;
+                // Verify that all temporal ranges of each backing index is still valid:
+                dataStream.validate(finalMBuilder::get);
```

Comment:
An exception here short-circuits the rest of the processing and fails the CS update, but I don't think we want to do that. I think if one data stream fails validation, we still want to try and update all the other data streams.

Should we catch the exception and log at the ERROR level here, but not skip the cluster state update? I'm concerned that if we don't then one failure means potentially no other data streams can be auto-adjusted.

Reply:
Good point. I don't think we want any data stream to be updated, because one data stream isn't valid any more. I will address this.